### PR TITLE
feat: Cloudformation logging

### PR DESCRIPTION
**Pull request by ammarkarachi**
*Issue #, if available:*

*Description of changes:*

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: packages/amplify-cli-logger/package.json [hunk 1]
```diff
@@ -0,0 +1,27 @@
...
+    "url": "git+https://github.com/aws-amplify/amplify-cli.git"
+  },
+  "author": "",
+  "license": "Apache-2.0",
+  "bugs": {
+    "url": "https://github.com/aws-amplify/amplify-cli/issues"
+  },
+  "homepage": "https://github.com/aws-amplify/amplify-cli#readme",
+  "dependencies": {
+    "winston": "^3.3.3",
```
Comment (undefobj) [inline packages/amplify-cli-logger/package.json:23]: I noticed Yathi used https://github.com/aws-amplify/amplify-cli/blob/f6dac5b6d55867e35b28b1c3eec9a6eeb4e4fbe3/packages/amplify-appsync-simulator/src/server/subscription/mqtt-server/client.ts#L7

Any pros/cons to this vs Winston?

File: packages/amplify-cli-logger/src/AmplifyLogger.ts [hunk 2]
```diff
@@ -0,0 +1,87 @@
...
+import { LocalProjectData, LogPayload, LogErrorPayload } from './Types';
+import { Redactor } from './Redactor';
+
+export class AmplifyLogger implements IAmplifyLogger {
+  logger: Logger;
+  format: winston.Logform.Format;
+  localProjectData!: LocalProjectData;
+  disabledAmplifyLogging: boolean = !!process.env.AMPLIFY_CLI_DISABLE_LOGGING;
+  options: any = {
+    mode: 0o640,
```
Comment (undefobj) [inline packages/amplify-cli-logger/src/AmplifyLogger.ts:16]: To help with future maintenance, when adding config in code that is opaque in nature (such as a hex value) is to add a comment on what this represents and *why* we are choosing this value. Can you add a note here?

File: packages/amplify-cli-logger/src/AmplifyLogger.ts [hunk 3]
```diff
@@ -0,0 +1,87 @@
...
+    if (!this.disabledAmplifyLogging) {
+      this.logger.add(
+        new winstonDailyRotateFile({
+          auditFile: getLogAuditFilePath(),
+          filename: getLogFilePath(),
+          datePattern: constants.DATE_PATTERN,
+          maxFiles: `${constants.MAX_FILE_DAYS}d`,
+          handleExceptions: false,
+          format: this.format,
+          //options: this.options,
```
Comment (undefobj) [inline packages/amplify-cli-logger/src/AmplifyLogger.ts:30]: Piggybacking on last comment, if we're not including this then we should delete both this comment as well as the config section above.

File: packages/amplify-cli-logger/src/AmplifyLogger.ts [hunk 4]
```diff
@@ -0,0 +1,87 @@
...
+    if (!this.disabledAmplifyLogging) {
+      this.logger.add(
+        new winstonDailyRotateFile({
+          auditFile: getLocalAuditLogFile(projecPath),
+          filename: getLocalLogFilePath(projecPath),
+          datePattern: constants.DATE_PATTERN,
+          maxFiles: `${constants.MAX_FILE_DAYS}d`,
+          handleExceptions: false,
+          format: this.format,
+          //options: this.options,
```
Comment (undefobj) [inline packages/amplify-cli-logger/src/AmplifyLogger.ts:68]: Same as above

File: packages/amplify-cli-logger/src/AmplifyLogger.ts [hunk 5]
```diff
@@ -0,0 +1,87 @@
...
+  constructor() {
+    this.logger = winston.createLogger();
+    this.format = format.combine(format.timestamp(), format.splat(), format.printf(this.formatter));
+    if (!this.disabledAmplifyLogging) {
+      this.logger.add(
+        new winstonDailyRotateFile({
+          auditFile: getLogAuditFilePath(),
+          filename: getLogFilePath(),
+          datePattern: constants.DATE_PATTERN,
+          maxFiles: `${constants.MAX_FILE_DAYS}d`,
```
Comment (undefobj) [inline packages/amplify-cli-logger/src/AmplifyLogger.ts:27]: A couple nits on the maintainability of this config object:
1. I try to not mix functions and constants. If you can export constants from your `getLogFilePath` above then it's better to stay consistent that way future refactors are done in one place.
2. Having a string literal along with a `d` formatter at the end is a problem with maintenance as well, especially when you've got this code lower on line 35. I can see a bug in the future where you change it in one place but not the other.

File: packages/amplify-cli-logger/src/AmplifyLogger.ts [hunk 6]
```diff
@@ -0,0 +1,87 @@
...
+      );
+    }
+  }
+
+  private formatter(info: winston.Logform.TransformableInfo): string {
+    const format = `${info.timestamp}|${info.level} : ${info.message}`;
+    if (info.level === 'info') {
+      if (info.isStackEvent) return format;
+      return `${format}(${info.args
+        .map((arg: any) => {
```
Comment (undefobj) [inline packages/amplify-cli-logger/src/AmplifyLogger.ts:41]: Can we make this `arg: string` and remove the need to stringify later?

File: packages/amplify-cli-logger/src/IAmplifyLogger.ts [hunk 7]
```diff
@@ -0,0 +1,8 @@
+import { LogErrorPayload, LogPayload } from './Types';
+
+export interface IAmplifyLogger {
+  log(message: string): void;
```
Comment (undefobj) [inline packages/amplify-cli-logger/src/IAmplifyLogger.ts:4]: How is `log` different from `logInfo`? Is there something more descriptive?

File: packages/amplify-cli-logger/src/Redactor.ts [hunk 8]
```diff
@@ -0,0 +1,59 @@
+const containsToRedact = ['key', 'id', 'password', 'name', 'arn', 'address', 'app', 'bucket'];
```
Comment (undefobj) [inline packages/amplify-cli-logger/src/Redactor.ts:1]: Nice 👍

File: packages/amplify-cli-logger/src/Redactor.ts [hunk 9]
```diff
@@ -0,0 +1,59 @@
...
+const quotes = '\\\\?"';
+const keyMatcher = `\\w*?(${containsToRedact.join('|')})\\w*?`;
+const completeMatch = `${quotes}(${keyMatcher})${quotes}:\\s?${quotes}([^!\\\\?"]+)${quotes}`;
+
+export function Redactor(arg: string | undefined): string {
+  if (!arg) return '';
+
+  // matches any json and gives values in json
```
Comment (undefobj) [inline packages/amplify-cli-logger/src/Redactor.ts:9]: Nice this is very good - When faced with complex function that is hard to read for future maintainers (which could be you), it's a good idea to give the function a more descriptive name or add a comment describing what is happening 💯

File: packages/amplify-cli-logger/src/constants.ts [hunk 10]
```diff
@@ -0,0 +1,9 @@
+export const constants = {
+  DOT_AMPLIFY: '.amplify',
```
Comment (undefobj) [inline packages/amplify-cli-logger/src/constants.ts:2]: Nice 👍

File: packages/amplify-provider-awscloudformation/src/system-config-manager.js [hunk 11]
```diff
@@ -114,25 +116,29 @@ async function getRoleCredentials(context, profileName, profileConfig) {
...
-      expiration: roleData.Credentials.Expiration,
+    const assumeRoleRequest = {
+      RoleArn: profileConfig.role_arn,
+      RoleSessionName: roleSessionName,
+      DurationSeconds: profileConfig.duration_seconds,
+      ExternalId: profileConfig.external_id,
+      SerialNumber: profileConfig.mfa_serial,
+      TokenCode: mfaTokenCode,
     };
+    const log = logger('getRoleCredentials.sts.assumeRole', [assumeRoleRequest]);
```
Comment (undefobj) [inline packages/amplify-provider-awscloudformation/src/system-config-manager.js:129]: This seems somewhat sensitive as well. Is it part of the redacted content?

File: packages/amplify-provider-awscloudformation/src/utils/aws-logger.js [hunk 12]
```diff
@@ -0,0 +1,39 @@
...
+    });
+  } else {
+    logInfo({
+      mod: `${file}.${crumb}`,
+      args,
+    });
+  }
+};
+
+export function logStackEvents(events) {
```
Comment (undefobj) [inline packages/amplify-provider-awscloudformation/src/utils/aws-logger.js:18]: Can we stay consistent in the function syntax we use here? Either everything should be arrow or nothing should be. All function prototypes should be of the same format if possible when defining new functionality.

**Review by undefobj — COMMENTED** [review]
Looks good, most of my comments are related to maintainability. I also see a `createService` refactor but I'm unsure what PR that happened in as it seems to be unrelated?

File: packages/amplify-cli-logger/package.json (hunk 1 above)
Comment (ammarkarachi) [reply packages/amplify-cli-logger/package.json:23]: It's a more mature library with more transports than pino on the other hand pino is smaller but the daily transport file is not that configurable.

**Comment by ammarkarachi** [comment]
> Looks good, most of my comments are related to maintainability. I also see a `createService` refactor but I'm unsure what PR that happened in as it seems to be unrelated?

I refactored that part as attaching the logger directly to the aws-sdk but it was not proving easy to redact information using. I left the refactor in since it was acting like a initializer with creds for all of the services

File: packages/amplify-cli-logger/src/IAmplifyLogger.ts (hunk 7 above)
Comment (ammarkarachi) [reply packages/amplify-cli-logger/src/IAmplifyLogger.ts]: it's left over from plugging in the logger directly to aws-sdk removed it now

File: packages/amplify-cli-logger/src/AmplifyLogger.ts (hunk 5 above)
Comment (ammarkarachi) [reply packages/amplify-cli-logger/src/AmplifyLogger.ts]: I am generating the folder name in the getLogFilePath. Depending on if it's `amplify-dev` or `amplify`

File: packages/amplify-provider-awscloudformation/src/system-config-manager.js (hunk 11 above)
Comment (ammarkarachi) [reply packages/amplify-provider-awscloudformation/src/system-config-manager.js:129]: Yeah I saw the arguments it uses and those values will be redacted

**Review by undefobj — APPROVED** [review]

**Pull request merged**